internal/caller/linux/linuxconfig.go

```diff
@@ -0,0 +1,136 @@
+package caller
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+)
+
+var (
+	installPath   = "/etc/wireguard"
+	tunnelTemplate = `[Interface]
+PrivateKey = %s
+Address = %s
+	
+[Peer]
+PublicKey = %s
+AllowedIPs %s
+Endpoint = %s`
+)
+
+// LinuxConfig represents the tunnel configuration for a linux machine.
+type LinuxConfig struct {
+	localDevice wgtypes.Device
+	tunnel      string
+	confName    string
+	allowedIps  string
+	addr        string
+	endpoint    string
+}
+
+// init transmits the linux config factory to the caller Client.
+func init() {
+	AppConfigurator = NewConfig()
+}
+
+// NewConfig generates and returns a new LinuxConfig.
+func NewConfig() *LinuxConfig {
+	newDevice := wgtypes.Device{
+		Name: "Linux Configurator",
+		Type: wgtypes.Userspace,
+	}
+
+	newLinuxConfig := LinuxConfig{
+		localDevice: newDevice,
+		tunnel:      tunnelTemplate,
+		confName:    "",
+	}
+	return &newLinuxConfig
+}
+
+// GenerateKeys creates a new set of Private and Public Keys for the
+// LinuxConfig.
+func (l *LinuxConfig) GenerateKeys() error {
+	privKey, err := wgtypes.GeneratePrivateKey()
+	if err != nil {
+		return err
+	}
+	l.localDevice.PrivateKey = privKey
+	l.localDevice.PublicKey = privKey.PublicKey()
+	return nil
+}
+
+// UpdateTunnel refreshes the LinuxConfig's tunnel with the latest values.
+func (l *LinuxConfig) UpdateTunnel() {
+	newTunnel := fmt.Sprintf(tunnelTemplate,
+		l.localDevice.PrivateKey,
+		l.addr,
+		l.localDevice.PublicKey,
+		l.allowedIps,
+		l.endpoint)
+	l.tunnel = newTunnel
+}
+
+// InstallConfig installs the wireguard tunnel to the machine.
+func (l *LinuxConfig) InstallConfig() error {
+
+	// Create tunnel
+	file := "./" + l.confName + ".conf"
+	f, err := os.Create(file)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = f.Write([]byte(l.tunnel))
+	if err != nil {
+		return err
+	}
+	_, err = os.Stat(installPath)
+	if err != nil {
+		cmd := exec.Command("sudo", "mkdir", installPath)
+		errN := cmd.Run()
+		if err != nil {
+			return errN
+		}
+	}
+
+	// Copy tunnel to /etc directory
+	wgConf := installPath + l.confName + ".conf"
+	cmd := exec.Command("sudo", "cp", file, wgConf)
+	err = cmd.Run()
+	if err != nil {
+		return err
+	}
+
+	// Start service
+	cmd = exec.Command("sudo", "wg-quick", "up", l.confName)
+	err = cmd.Run()
+	if err != nil {
+		return err
+	}
+	_ = os.Remove(file)
+	return nil
+}
+
+// UninstallConfig removes the Wireguard tunnel from the machine.
+func (l *LinuxConfig) UninstallConfig() error {
```

Comment:
Similarly, this function needs to be rewritten to not do the things that its trying to do here.  See comments above.